### PR TITLE
Chore: Update rules to use the new PassesTest method instead of Evaluate

### PR DESCRIPTION
**Pull request by RobGallo**
#### Describe the change

    Adding the evaluation code to the rule's info will allow us to automatically generate documentation which provides the rule's expected error code.

This is a large change. When reviewing please try to make sure the expected error code for each rule is consistent with its former value. (Those changes were made by hand.)

All the changes in the test files were made by a search and replace.

### CODE REVIEW

File: src/Rules/Library/IsContentElementTrueOptional.cs [hunk 1]
```diff
@@ -20,11 +20,12 @@ public IsContentElementTrueOptional()
...
             this.Info.Standard = A11yCriteriaId.ObjectInformation;
             this.Info.PropertyID = PropertyType.UIA_IsContentElementPropertyId;
+            this.Info.ErrorCode = EvaluationCode.Open;
         }
 
-        public override EvaluationCode Evaluate(IA11yElement e)
+        public override bool PassesTest(IA11yElement e)
         {
-            return EvaluationCode.Open;
+            return true;
```
Comment (DaveTryon) [inline src/Rules/Library/IsContentElementTrueOptional.cs:28]: Won't this result in an evaluation code of Pass instead of Open?

File: src/Rules/Library/ListItemSiblingsUnique.cs [hunk 2]
```diff
@@ -32,9 +34,9 @@ public override EvaluationCode Evaluate(IA11yElement e)
                 & Name.Is(e.Name)
                 & LocalizedControlType.Is(e.LocalizedControlType));
             var count = siblings.GetValue(e);
-            if (count < 1) return EvaluationCode.RuleExecutionError;
+            if (count < 1) throw new Exception(String.Format(CultureInfo.CurrentCulture, ErrorMessages.NoElementFound, this.Info.ID));
```
Comment (DaveTryon) [inline src/Rules/Library/ListItemSiblingsUnique.cs:37]: Not entirely sure why the Exception is different, but I'd suggest CultureInfo.InvariantCulture instead of CultureInfo.CurrentCulture (the message isn't localized, so it doesn't make sense to localize the ID)

File: src/Rules/Library/LocalizedControlTypeIsNotEmpty.cs [hunk 3]
```diff
@@ -19,13 +19,15 @@ public LocalizedControlTypeIsNotEmpty()
             this.Info.HowToFix = HowToFix.LocalizedControlTypeNotEmpty;
             this.Info.Standard = A11yCriteriaId.ObjectInformation;
             this.Info.PropertyID = PropertyType.UIA_LocalizedControlTypePropertyId;
+
```
Comment (DaveTryon) [inline src/Rules/Library/LocalizedControlTypeIsNotEmpty.cs:22]: Remove extra newline on line 22 (you didn't add it in other classs so I'm assuming this is an oversight)

File: src/Rules/Library/NameIsEmptyButElementIsNotKeyboardFocusable.cs [hunk 4]
```diff
@@ -21,13 +22,14 @@ public NameIsEmptyButElementIsNotKeyboardFocusable()
...
             this.Info.Standard = A11yCriteriaId.ObjectInformation;
             this.Info.PropertyID = PropertyType.UIA_NamePropertyId;
+            this.Info.ErrorCode = EvaluationCode.Open;
         }
 
-        public override EvaluationCode Evaluate(IA11yElement e)
+        public override bool PassesTest(IA11yElement e)
         {
-            if (e == null) return EvaluationCode.RuleExecutionError;
+            if (e == null) throw new ArgumentNullException(nameof(e));
```
Comment (DaveTryon) [inline src/Rules/Library/NameIsEmptyButElementIsNotKeyboardFocusable.cs:30]: This Exception makes sense, but there are lots of other rules that will merrily throw a NulReferenceException in this case (for example, NameIsInformative (chosen just because it's the next in the review) has no guard. It feels like the guard ought to be consistent in all rules (could be a separate PR)

File: src/Rules/Library/ParentChildShouldNotHaveSameNameAndLocalizedControlType.cs [hunk 5]
```diff
@@ -18,14 +19,15 @@ public ParentChildShouldNotHaveSameNameAndLocalizedControlType()
...
-        public override EvaluationCode Evaluate(IA11yElement e)
+        public override bool PassesTest(IA11yElement e)
         {
-            if (e == null) return EvaluationCode.RuleExecutionError;
-            if (e.Parent == null) return EvaluationCode.RuleExecutionError;
+            if (e == null) throw new ArgumentNullException(nameof(e));
+            if (e.Parent == null) throw new ArgumentException(ErrorMessages.ElementParentNull, nameof(e));
 
-            return e.Name == e.Parent.Name && e.LocalizedControlType == e.Parent.LocalizedControlType ? EvaluationCode.Error : EvaluationCode.Pass;
+            return e.Name != e.Parent.Name || e.LocalizedControlType != e.Parent.LocalizedControlType;
```
Comment (DaveTryon) [inline src/Rules/Library/ParentChildShouldNotHaveSameNameAndLocalizedControlType.cs:30]: I wish we had some parentheses here, just to be explicit about grouping:
```suggestion
            return (e.Name != e.Parent.Name) || (e.LocalizedControlType != e.Parent.LocalizedControlType);
```

File: src/Rules/Library/SiblingUniqueAndFocusable.cs [hunk 6]
```diff
@@ -48,9 +50,9 @@ public override EvaluationCode Evaluate(IA11yElement e)
                 & Name.Is(e.Name)
                 & LocalizedControlType.Is(e.LocalizedControlType));
             var count = siblings.GetValue(e);
-            if (count < 1) return EvaluationCode.RuleExecutionError;
+            if (count < 1) throw new Exception(string.Format(CultureInfo.CurrentCulture, ErrorMessages.NoElementFound, this.Info.ID));
```
Comment (DaveTryon) [inline src/Rules/Library/SiblingUniqueAndFocusable.cs:53]: Not entirely sure why the Exception is different, but I'd suggest CultureInfo.InvariantCulture instead of CultureInfo.CurrentCulture (the message isn't localized, so it doesn't make sense to localize the ID)
```suggestion
            if (count < 1) throw new Exception(string.Format(CultureInfo.InvariantCulture, ErrorMessages.NoElementFound, this.Info.ID));
```

File: src/Rules/Library/SiblingUniqueAndNotFocusable.cs [hunk 7]
```diff
@@ -47,9 +49,9 @@ public override EvaluationCode Evaluate(IA11yElement e)
                 & Name.Is(e.Name)
                 & LocalizedControlType.Is(e.LocalizedControlType));
             var count = siblings.GetValue(e);
-            if (count < 1) return EvaluationCode.RuleExecutionError;
+            if (count < 1) throw new Exception(string.Format(CultureInfo.CurrentCulture, ErrorMessages.NoElementFound, this.Info.ID));
```
Comment (DaveTryon) [inline src/Rules/Library/SiblingUniqueAndNotFocusable.cs:52]: Not entirely sure why the Exception is different, but I'd suggest CultureInfo.InvariantCulture instead of CultureInfo.CurrentCulture (the message isn't localized, so it doesn't make sense to localize the ID)
```suggestion
            if (count < 1) throw new Exception(string.Format(CultureInfo.InvariantCulture, ErrorMessages.NoElementFound, this.Info.ID));
```

**Review by DaveTryon — CHANGES_REQUESTED** [review]
I've only reviewed the main classes so far (no unit tests), but since I found issues there, I'll review the unit tests on the next iteration

File: src/Rules/Resources/ErrorMessages.resx [hunk 8]
```diff
@@ -150,6 +150,9 @@
   <data name="NoControlTypeEntryFound" xml:space="preserve">
     <value>No control type entry was found in dictionary</value>
   </data>
+  <data name="NoElementFound" xml:space="preserve">
+    <value>Based on the Rule.Condition match, there should have been at least one element found in the PassesTest method of {0}</value>
```
Comment (DaveTryon) [inline src/Rules/Resources/ErrorMessages.resx:154]: Will this be exposed to users? It seems a little too obscure to really help them

File: src/RulesTest/Library/NameIsEmptyButElementIsNotKeyboardFocusableTest.cs [hunk 9]
```diff
@@ -1,8 +1,8 @@
 // Copyright (c) Microsoft. All rights reserved.
 // Licensed under the MIT license. See LICENSE file in the project root for full license information.
 using Microsoft.VisualStudio.TestTools.UnitTesting;
+using System;
```
Comment (DaveTryon) [inline src/RulesTest/Library/NameIsEmptyButElementIsNotKeyboardFocusableTest.cs:4]: In "using System;" added intentionally? No changes in this file seem to necessitate it

File: src/Rules/Library/IsContentElementTrueOptional.cs (hunk 1 above)
Comment (RobGallo) [reply src/Rules/Library/IsContentElementTrueOptional.cs:28]: I suspected I might make that mistake at some point. Nice catch.

File: src/Rules/Library/ListItemSiblingsUnique.cs (hunk 2 above)
Comment (RobGallo) [reply src/Rules/Library/ListItemSiblingsUnique.cs:37]: Why the exception is different from what?

And I am going to fix these instances in a single change so I can make sure they compile.

File: src/Rules/Library/NameIsEmptyButElementIsNotKeyboardFocusable.cs (hunk 4 above)
Comment (RobGallo) [reply src/Rules/Library/NameIsEmptyButElementIsNotKeyboardFocusable.cs:30]: Agreed.

File: src/Rules/Library/SiblingUniqueAndFocusable.cs (hunk 6 above)
Comment (RobGallo) [reply src/Rules/Library/SiblingUniqueAndFocusable.cs:53]: Fixed.

File: src/Rules/Library/SiblingUniqueAndNotFocusable.cs (hunk 7 above)
Comment (RobGallo) [reply src/Rules/Library/SiblingUniqueAndNotFocusable.cs:52]: Fixed.

File: src/Rules/Library/ListItemSiblingsUnique.cs (hunk 2 above)
Comment (DaveTryon) [reply src/Rules/Library/ListItemSiblingsUnique.cs]: I understand the Exception now--you're throwing here and RuleExecutionError will come from the Exception handler.

**Review by DaveTryon — APPROVED** [review]
